curp/src/server/curp_node.rs

```diff
@@ -173,42 +176,109 @@ impl<C: 'static + Command> CurpNode<C> {
         loop {
             let _now = ticker.tick().await;
-            let action = curp.tick();
-            match action {
-                TickAction::Heartbeat(hbs) => {
-                    Self::bcast_heartbeats(Arc::clone(&curp), &connects, hbs).await;
-                }
-                TickAction::Votes(votes) => {
-                    Self::bcast_votes(Arc::clone(&curp), &connects, votes).await;
-                }
-                TickAction::Nothing => {}
+            if let Some(vote) = curp.tick_election() {
+                Self::bcast_vote(curp.as_ref(), &connects, vote).await;
             }
         }
     }
 
-    /// Background leader calibrate followers
-    async fn calibrate_task(
+    /// Sync task is responsible for syncing a follower
+    async fn sync_task(
         curp: Arc<RawCurp<C>>,
-        connects: HashMap<ServerId, Arc<impl ConnectApi>>,
-        mut calibrate_rx: mpsc::UnboundedReceiver<ServerId>,
+        connect: Arc<impl ConnectApi>,
+        sync_event: Arc<Event>,
+    ) {
+        let leader_event = curp.leader_event();
+        loop {
+            if !curp.is_leader() {
+                leader_event.listen().await;
+            }
+            Self::_sync_task(
+                Arc::clone(&curp),
+                Arc::clone(&connect),
+                Arc::clone(&sync_event),
+            )
+            .await;
+        }
+    }
+
+    /// real sync task, will stop if self is no long the leader
+    async fn _sync_task(
+        curp: Arc<RawCurp<C>>,
+        connect: Arc<impl ConnectApi>,
+        sync_event: Arc<Event>,
     ) {
-        let mut handlers: HashMap<ServerId, JoinHandle<()>> = HashMap::new();
-        while let Some(follower_id) = calibrate_rx.recv().await {
-            if handlers
-                .get(&follower_id)
-                .map_or(false, |hd| !hd.is_finished())
-            {
+        let mut hb_opt = false;
+        let mut ticker = tokio::time::interval(curp.cfg().heartbeat_interval);
+        ticker.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        let id = connect.id();
+        let rpc_timeout = curp.cfg().rpc_timeout;
+
+        let mut listener = None;
+
+        #[allow(clippy::integer_arithmetic, clippy::unwrap_used)]
+        // tokio select internal triggered,
+        loop {
+            if listener.is_none() {
+                listener = Some(sync_event.listen());
+            }
+            // a sync is either triggered by an heartbeat timeout event or when new log entries arrive
+            tokio::select! {
+                _now = ticker.tick() => {
+                    if hb_opt {
+                        hb_opt = false;
+                        continue;
+                    }
+                }
+                // TODO: use a batch timer
+                _ = listener.take().unwrap() => {
+                }
+            }
+            listener = Some(sync_event.listen());
```

Comment:
Lines 239 and 224-226 seem to be doing the same thing, I think we can remove line 239, or we can remove line 224-226 and init listener at line 219

Reply:
Fixed